src/compiler/checker.ts

```diff
@@ -6753,9 +6763,132 @@ namespace ts {
+                let results: Signature[] | undefined = masterList.slice();
+                for (const signatures of signatureLists) {
+                    if (signatures === masterList) continue;
+                    const signature = signatures[0];
+                    Debug.assert(!!signature, "getUnionSignatures bails early on empty signature lists and should not have empty lists on second pass");
+                    results = mergeAsUnionSignature(results, signature);
+                    if (!results) {
+                        break;
+                    }
+                }
+                result = results;
+            }
             return result || emptyArray;
         }
 
+        function mergeAsUnionSignature(masterList: Signature[], addition: Signature): Signature[] | undefined {
+            if (addition.typeParameters && some(masterList, s => !!s.typeParameters)) {
+                return; // Can't currently unify type parameters
+            }
+            return map(masterList, sig => combineSignaturesOfUnionMembers(sig, addition));
+        }
+
+        function combineUnionThisParam(left: Symbol | undefined, right: Symbol | undefined): Symbol | undefined {
+            if (!left && !right) {
+                return;
+            }
+            if (!left) {
+                return right;
+            }
+            if (!right) {
+                return left;
+            }
+            // A signature `this` type might be a read or a write position... It's very possible that it should be invariant
+            // and we should refuse to merge signatures if there are `this` types and they do not match. However, so as to be
+            // permissive when calling, for now, we'll union the `this` types just like the overlapping-union-signature check does
+            const thisType = getUnionType([getTypeOfSymbol(left), getTypeOfSymbol(right)], UnionReduction.Subtype);
+            return createSymbolWithType(left, thisType);
+        }
+
+        function combinePartameterNames(index: number, left: __String | undefined, right: __String | undefined): __String {
+            if (!left && !right) {
+                return `arg${index}` as __String;
+            }
+            if (!left) {
+                return right!;
+            }
+            if (!right || left === right) {
+                return left;
+            }
+            const names = createMap<true>();
+            // We join names in an upperCamelCase fashion.
+            const seperator = "And";
+            const leftNames = unescapeLeadingUnderscores(left).split(seperator);
+            const rightNames = unescapeLeadingUnderscores(right).split(seperator);
+            let first = true;
+            for (const list of [leftNames, rightNames]) {
+                for (const n of list) {
+                    names.set(first ? (first = false, n) : (n.slice(0, 1).toUpperCase() + n.slice(1)), true);
+                }
+            }

```

Comment:
I'm not sure how well this is going to work for foreign languages!

Reply:
Poorly! 😄